Questions 1
1. (2 points) Assume that you are buying and selling electricity. The more electricity you buy, the better 
price you get per kwH and the buying price of electricity follows function f(x)=1−0.01x, where 0 ≤ x ≤ 
50 is the amount of electricity you buy. On the other hand, the price that you get from selling electricity 
follows the function g(x)=2−0.01x^2 with x again the amount of electricity that you sell. Formulate an 
optimization problem that maximizes the profit.

Solution 1

#Constriants:   unit of electricity = x, 

                buying price
                f(x) = 1 - 0.01x, 
                
                selling  price
                g(x) = 2 - 0.01x^2

P(x) = x (g(x) - f(x))

     = x (2 - 0.01x^2 - 1 + 0.01x)
     
     = x (1 - 0.01x^2 + 0.01x)

max P(x) = x (1 - 0.01x^2 + 0.01x)

s.t. 0 <= x <= 50


Question 2

2. (2 points) Use bisection search to optimize the problem
min (1 − x)^2 + x
st x ∈ [0,2]
Code the search algorithm by yourself, not use the one shown in the lecture material.

Solution 2

The code shown in the lecture material compares the middle point m at f(m + epsilon) > f(m – epsilon) to 
decide the function monotonicity. It is difficult to be creative about simple methods, yet, I used the slope of function 
to decide the function monotonicity which I believe share the same idea.

Python code  for Question 2

We begin by defining the function f(x);
Next, we find the function monotonicity, when increasing implies the slope will be greater than 0
Also, the bisection logic means if function f(x) at middle point mid is increasing, then we cut the end point to 
the middle, if it is decreasing, then we let the start point be the middle point.

#This method it can only be used to find local minimum, if you want to find the maximum, you 
need to let the start point be the middle point if it is increasing. 

Using loop and recursion method to implement bisection as presented in the following code:

In [5]:
def f(x):
    return (1.0-x)**2 + x

def monotonicity(f, x1, delta):
    #lim([f(x1) - f(x1 - delta)]/ delta) when delta -> 0,
    #if the limit is greater than 0, then, the function is increasing,
    y = (f(x1) - f(x1-delta))/delta
    if y > 0: return 1          #function increasing
    elif y == 0: return 0       #it is not likely function will find extreme point because delta! = 0
    else: return -1             #function decreasing

"""In order to find the optimal,
we set the start point of the function as parameter: a
we set the end point of the function as parameter: b """

def find_min_bisection(a, b, f, delta):
    start = a
    end = b
    while start < end:
        mid = (start+end)/2
        increasing = monotonicity(f, mid, delta)
        if increasing == 1:
            end = mid+delta
        elif increasing == -1:
            start = mid-delta
        else:
            return mid
        return (start+end)/2
    
def recursion_bisection_min(a, b, f, delta):
    start = a
    end = b
    mid = (start+end)/2
    increasing = monotonicity(f, mid, delta)

    if increasing == 1:
        end = mid+delta
        return recursion_bisection_min(start, end, f, delta)
    elif increasing == -1:
        start = mid-delta
        return recursion_bisection_min(start, end, f, delta)
    else:
        return mid
    
x = recursion_bisection_min(0.00, 5.0, f, 0.001);
print("optimum is " + str(x) +", function value is " + str(f(x)))

optimum is 0.5004999999999953, function value is 0.75000025


Test Result for Question 2

The result meets the math validation.
optimum is 0.5004999999999953, function value is 0.75000025

Question 3

3. (2 points) Use the golden section search to optimize the problem from the above exercise. Code the 
search algorithm by yourself, not use the one shown in the lecture material.

### Solution 3

The changes is using slope to decide the monotonicity by using slop function 
k = (y2 − y1)/(x2 − x1)

The first golden point d (x2,y2), the second golden point near the start point 'a' will be c(x1,y1);
and we know x1 and x2 by using the proportional function 
x1 = a + ((√5 − 1) / 2) * (b - a)       
                                        #where a and b is the chosen point
x2 = b − ((√5 − 1) / 2) * (b - a)


In [6]:
import math

R = (math.sqrt(5)-1) * 0.5

def f(x):
    return (1.0-x) ** 2 + x

def monotonicity(x1, y1, x2, y2):
    slope = (y2-y1)/(x2-x1)
    if slope > 0: return 1
    elif slope < 0: return -1
    else: return 0

def golden_section_min(a, b, f, delta):
    start = a
    end = b
    while start < end:
        d = start+R * (end-start)
        c = end-R * (end-start)
        increasing = monotonicity(c, f(c), d, f(d))
        if increasing == 1:
            end = d
        elif increasing == -1:
            start = c
        else:
            return d
        return start+R * (end-start)

def g(x):
    return (x-10) ** 2+10

x = golden_section_min(0.0, 24.0, f, 0.001);
print("optimum is " + str(x) +", function value is " + str(f(x)))

optimum is 9.167184270002526, function value is 75.87008317017921


This is result of function f,
optimum is 9.167184270002526, function value is 75.87008317017921

### Question 4

<p>
(2 points) Use differentiation/derivation to optimize the optimization problem of task 1 (above) 
and to verify the answer of the optimization problem in task 2 (above).
</p>

### Solution 4

$$
\begin{align}
4.1 \quad the \quad optimization \quad problem \quad of \quad task \quad 1\\

max P(x) = x (1 - 0.01x^2 + 0.01x)\\
s.t. 0 <= x <= 50\\
P'(x) = f'(x)g(x) + f(x)g'(x) = 1 * (1 - 0.001x^2 + 0.01x) + x * (0.01 - 0.02x)\\
P'(x) = -0.01x^2 + x(0.01 - 0.02x) + 0.01x + 1\\
P'(x) = -0.03x^2 + 0.02x + 1\\
P"(x) = 0.02 - 0.06x\\

∴ when \quad P'(x) = 0, \quad at \quad x = 1/2 \quad had \quad a \quad maximal \quad P(x) = 21/5\\
\\

4.2 Optimization \quad problem \quad in \quad task \quad 2\\

min f(x) = (1 - x)^2 + x\\
st  x ∈ [0, 2]\\
f'(x) = 2x - 1\\
f"(x) = 2 > 0\\
when f'(x) = 0, \quad x = 1/2 \quad had \quad a  \quad minimal \quad f(x) = 3/4.\\
\end{align}